### PR TITLE
Contribute DiagnosticCountInPassHook

### PR DESCRIPTION
**Pull request by vlstill**
This is the reason behind #4626. A pass manager hook that is useful for testing (p4test-like) the compiler output or just for the developers to see what passes emit diagnostic messages.

Example output (with the hook in driver added and running with `-TdiagnosticCountInPass:1`):

```
main.p4(65): [--Wwarn=parser-transition] warning: sub_parser: implicit transition to `reject'
           state sub_parser{
                 ^^^^^^^^^^
PASS FrontEnd ~> CreateBuiltins emitted 1 warning
[--Wwarn=parser-transition] warning: SelectCase: unreachable case
PASS MidEnd ~> PassRepeated emitted 1 warning
```

### CODE REVIEW

File: ir/pass_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
+#include "ir/pass_utils.h"
+
+#include <initializer_list>
+#include <tuple>
+
+#include "lib/log.h"
+
+/// @file
+/// @authors Vladimír Štill
```
Comment (fruffy) [inline ir/pass_utils.cpp:9]: Nit: A little non-standard, should we use a copyright notice instead?

File: ir/pass_utils.h [hunk 2]
```diff
@@ -0,0 +1,35 @@
+#ifndef IR_PASS_UTILS_H_
+#define IR_PASS_UTILS_H_
+
+#include "ir/pass_manager.h"
+#include "lib/compile_context.h"
+
+/// @file
```
Comment (fruffy) [inline ir/pass_utils.h:7]: Same here.

File: ir/pass_utils.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+                  std::tuple(warns, state->lastWarningCount, "warning"),
+                  std::tuple(infos, state->lastInfoCount, "info")}) {
+                if (newCnt > oldCnt) {
+                    auto diff = newCnt - oldCnt;
+                    summary << sep << diff << " " << kind << (diff > 1 ? "s" : "");
+                    sep = ", ";
+                }
+            }
+
+            // if the log level is high enough, emit also pass count
```
Comment (fruffy) [inline ir/pass_utils.cpp:47]:
```suggestion
            // If the log level is high enough, emit also pass count.
```

**Review by fruffy — APPROVED** [review]
Could be nice to have an example exercising this hook.

File: ir/pass_utils.cpp (hunk 1 above)
Comment (vlstill) [reply ir/pass_utils.cpp:9]: Is there a "standard" copyright notice in modern "P4C"? I'm mostly finding old barefoot notices...

Or I can just skip this.

**Comment by vlstill** [comment]
> Could be nice to have an example exercising this hook.

Let me at it to P4test, the output has to be enabled anyway so it should not break anything.

**Comment by vlstill** [comment]
```bash
$ ./p4c/p4test -TdiagnosticCountInPass:1 <(printf 'const bit<42> x = -4;\nconst bit<16> y = (bit<16>)x;\nvoid foo() { return 4; }')
/dev/fd/63(1): [--Wwarn=mismatch] warning: -42w4: negative value with unsigned type
const bit<42> x = -4;
                   ^
/dev/fd/63(1): [--Wwarn=mismatch] warning: 16w4398046511100: value does not fit in 16 bits
const bit<42> x = -4;
                   ^
PASS ConstantFolding ~> DoConstantFolding emitted 2 warnings
/dev/fd/63(3): [--Werror=type-error] error: return 4: return expression in function with void return
void foo() { return 4; }
             ^^^^^^
PASS FrontEnd ~> P4::TypeInference emitted 1 error

**Review by fruffy — APPROVED** [review]

File: ir/pass_utils.cpp (hunk 1 above)
Comment (fruffy) [reply ir/pass_utils.cpp]: I do not think there is a standard, but the Apache template we use is the most common one. Recent example here: https://github.com/p4lang/p4c/blob/main/midend/def_use.h#L2 

But we also have https://github.com/p4lang/p4c/blob/e30416337cc0e3bf046909324aa7ae0b160f5224/lib/nethash.cpp#L23 here now, so I think we can go with either version.

**Comment by vlstill** [comment]
Any idea what is wrong with the Fedora build?

>  /__w/p4c/p4c/build/_deps/abseil-src/absl/numeric/internal/bits.h:309:24: error: '__builtin_ctzs' needs isa option -mbmi

**Comment by fruffy** [comment]
> Any idea what is wrong with the Fedora build?
> 
> > /__w/p4c/p4c/build/_deps/abseil-src/absl/numeric/internal/bits.h:309:24: error: '__builtin_ctzs' needs isa option -mbmi

It looks like Fedora GCC switched from 13 to 14 and this instruction now needs an option.

**Pull request merged**